### PR TITLE
Add more positions

### PR DESCRIPTION
**Pull request by GiudGiud**
Planning
- [x] hex grid
- [x] incomplete hex grid
- [x] cartesian grid
- [x] nested positions
- [x] transform positions

I ended up fighting the system a bit to get the highlight done. With these it will be a lot smoother.
Transforms, nested and hexes should be usable by @jortensi on SFR / MR problems rn
Cartesian is future proofing

### CODE REVIEW

File: framework/include/positions/DistributedPositions.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+// Moose includes
+#include "Positions.h"
+
+/**
+ * Positions created by distributing Positions objects onto on another. The first positions object
```
Comment (lindsayad) [inline framework/include/positions/DistributedPositions.h:16]:
```suggestion
 * Positions created by distributing Positions objects onto one another. The first positions object
```

File: framework/include/positions/DistributedPositions.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * create the base positions onto which the later ones are translated from.
```
Comment (lindsayad) [inline framework/include/positions/DistributedPositions.h:17]:
```suggestion
 * creates the base positions onto which the later ones are translated from.
```

File: framework/include/positions/DistributedPositions.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ */
+class DistributedPositions : public Positions
+{
+public:
+  static InputParameters validParams();
+  DistributedPositions(const InputParameters & parameters);
+
+  void initialize() override;
```
Comment (lindsayad) [inline framework/include/positions/DistributedPositions.h:25]:
```suggestion
  virtual void initialize() override;
```

File: framework/include/positions/TransformedPositions.h [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * Positions transformed using a simple operation (translation, rotation, scaling)
+ */
+class TransformedPositions : public Positions
+{
+public:
+  static InputParameters validParams();
+  TransformedPositions(const InputParameters & parameters);
+  virtual ~TransformedPositions() = default;
+
+  void initialize() override;
```
Comment (lindsayad) [inline framework/include/positions/TransformedPositions.h:25]:
```suggestion
  virtual void initialize() override;
```

File: framework/include/positions/TransformedPositions.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+private:
+  /// Position object providing the transformed positions
+  const Positions * _base_positions;
+
+  /// Transformation to perform
+  const MooseEnum _transform;
```
Comment (lindsayad) [inline framework/include/positions/TransformedPositions.h:32]:
```suggestion
  const MooseEnum & _transform;
```

File: framework/src/positions/DistributedPositions.C [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+  params.addClassDescription(
+      "Distribute positions, using translations, over one or more positions");
+  return params;
+}
+
+DistributedPositions::DistributedPositions(const InputParameters & parameters)
+  : Positions(parameters)
+{
+
```
Comment (lindsayad) [inline framework/src/positions/DistributedPositions.C:39]:
```suggestion
```

File: framework/src/positions/DistributedPositions.C [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  const auto base_names = getParam<std::vector<PositionsName>>("positions");
```
Comment (lindsayad) [inline framework/src/positions/DistributedPositions.C:40]:
```suggestion
  const auto & base_names = getParam<std::vector<PositionsName>>("positions");
```

File: framework/src/positions/DistributedPositions.C [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+void
+DistributedPositions::initialize()
+{
+  clearPositions();
+  const bool initial = _fe_problem.getCurrentExecuteOnFlag() == EXEC_INITIAL;
+
+  // Check that everything is initialized
+  for (const auto pos_obj : _positions_objs)
```
Comment (lindsayad) [inline framework/src/positions/DistributedPositions.C:63]:
```suggestion
  for (const auto * const pos_obj : _positions_objs)
```

File: framework/src/positions/DistributedPositions.C [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    if (!pos_obj->initialized(initial))
+      mooseError("Positions '", pos_obj->name(), "' is not initialized.");
+
+  // Size new positions vector
+  unsigned int n_positions = 1;
+  for (const auto pos_obj : _positions_objs)
```
Comment (lindsayad) [inline framework/src/positions/DistributedPositions.C:69]:
```suggestion
  for (const auto * const pos_obj : _positions_objs)
```

File: framework/src/positions/DistributedPositions.C [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    n_positions *= pos_obj->getNumPositions(initial);
+  }
+  _positions.resize(n_positions);
+
+  // Fill _positions by distributing using translations
+  _positions[0] = Point(0, 0, 0);
+  unsigned int current_index = 1;
+  for (const auto pos_i : index_range(_positions_objs))
+  {
+    const auto current_positions = _positions_objs[_positions_objs.size() - 1 - pos_i];
```
Comment (lindsayad) [inline framework/src/positions/DistributedPositions.C:83]:
```suggestion
    const auto * const current_positions = _positions_objs[_positions_objs.size() - 1 - pos_i];
```

File: framework/src/positions/TransformedPositions.C [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  return params;
+}
+
+TransformedPositions::TransformedPositions(const InputParameters & parameters)
+  : Positions(parameters),
+    _transform(getParam<MooseEnum>("transform")),
+    _vector_value(getParam<RealVectorValue>("vector_value"))
+{
+
+  const auto base_name = getParam<PositionsName>("base_positions");
```
Comment (lindsayad) [inline framework/src/positions/TransformedPositions.C:53]:
```suggestion
  const auto & base_name = getParam<PositionsName>("base_positions");
```

File: framework/src/positions/TransformedPositions.C [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+  const bool initial = _fe_problem.getCurrentExecuteOnFlag() == EXEC_INITIAL;
+  if (!_base_positions->initialized(initial))
+    mooseError("Positions '", _base_positions->name(), "' is not initialized.");
+
+  const auto n_positions = _base_positions->getNumPositions(initial);
+  _positions.resize(n_positions);
+
+  for (const auto i : make_range(n_positions))
+  {
+    const auto base_point = _base_positions->getPosition(i, initial);
```
Comment (lindsayad) [inline framework/src/positions/TransformedPositions.C:79]:
```suggestion
    const auto & base_point = _base_positions->getPosition(i, initial);
```

File: framework/src/positions/TransformedPositions.C [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+  params.addRequiredParam<MooseEnum>(
+      "transform",
+      transforms,
+      "The type of transformation to perform (TRANSLATE, ROTATE_XYZ, SCALE)");
+  params.addParam<RealVectorValue>(
+      "vector_value",
+      "The value to use for the transformation. When using TRANSLATE or SCALE, the "
+      "xyz coordinates are applied in each direction respectively. When using "
+      "ROTATE_XYZ, the values are interpreted as rotations around the X, Y and Z axis, in that "
+      "order.");
```
Comment (lindsayad) [inline framework/src/positions/TransformedPositions.C:33]: Mention that this should be in degrees

File: test/tests/positions/transformed_positions.i [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+  []
+  [scale]
+    type = TransformedPositions
+    base_positions = 'input'
+    vector_value = '1 2 3'
+    transform = 'SCALE'
+  []
+  [rotate]
+    type = TransformedPositions
+    base_positions = 'input'
```
Comment (lindsayad) [inline test/tests/positions/transformed_positions.i:25]: I assume that you could also supply `scale` here?

File: test/tests/positions/transformed_positions.i [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    vector_value = '0 0 90'
+    transform = 'ROTATE_XYZ'
+  []
+  [translate]
+    type = TransformedPositions
+    base_positions = 'input'
```
Comment (lindsayad) [inline test/tests/positions/transformed_positions.i:31]: I assume that you could also supply `scale` or `rotate` either?

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: framework/src/positions/TransformedPositions.C [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+  MooseEnum transforms("TRANSLATE ROTATE_XYZ SCALE");
+  params.addRequiredParam<MooseEnum>(
+      "transform",
+      transforms,
+      "The type of transformation to perform (TRANSLATE, ROTATE_XYZ, SCALE)");
+  params.addParam<RealVectorValue>(
+      "vector_value",
+      "The value to use for the transformation. When using TRANSLATE or SCALE, the "
+      "xyz coordinates are applied in each direction respectively. When using "
+      "ROTATE_XYZ, the values are interpreted as rotations around the X, Y and Z axis, in that "
```
Comment (GiudGiud) [inline framework/src/positions/TransformedPositions.C:32]:
```suggestion
      "ROTATE_XYZ, the values are interpreted as rotations in degrees around the X, Y and Z axis, in that "
```

File: test/tests/positions/transformed_positions.i (hunk 14 above)
Comment (GiudGiud) [reply test/tests/positions/transformed_positions.i:25]: yeah if you wanted to chain them.

File: test/tests/positions/transformed_positions.i (hunk 15 above)
Comment (GiudGiud) [reply test/tests/positions/transformed_positions.i:31]: I ll try. This would be legitimate use

Comment (GiudGiud) [reply test/tests/positions/transformed_positions.i:31]: works as expected. 
Anything defined first is fine until we do dependency resolution at construction

File: framework/src/positions/TransformedPositions.C (hunk 13 above)
Comment (GiudGiud) [reply framework/src/positions/TransformedPositions.C]: done

**Review by lindsayad — APPROVED** [review]

**Comment by GiudGiud** [comment]
Thanks for the review!

**Comment by lewisgross1296** [comment]
Thanks for adding this in! I'll clone`idaholab/moose:next`and give it a go.

**Pull request merged**